code/obj/machinery/manufacturer.dm

```diff
@@ -1607,43 +1609,46 @@ TYPEINFO(/obj/machinery/manufacturer)
 				return P
 		return null
 
-	/// Returns associative list of manufacturing requirement to material piece reference, but does not guarantee all item_paths are satisfied or that
-	/// the blueprint will have the required materials ready by the time it reaches the front of the queue. Mats not used are not added to the return value
+	/// Returns associative list of manufacturing requirement to material piece references, but does not guarantee all item_paths are satisfied or that
+	/// the blueprint will have the required materials ready by the time it reaches the front of the queue. Reqs not satisfied are not added to mats_used
 	proc/get_materials_needed(datum/manufacture/M)
-		var/list/C = src.get_contents()
-
 		var/list/mats_used = list()
-		var/list/mats_projected = list()
-		for (var/obj/item/material_piece/P in C)
-			mats_projected["\ref[P]"] = P.amount * 10
+		var/list/mats_reserved = list()
 
 		for (var/datum/manufacturing_requirement/R as anything in M.item_requirements)
-			var/required_amount = M.item_requirements[R]
-			for (var/obj/item/material_piece/P in C)
-				var/P_ref = "\ref[P]"
-				if (mats_projected[P_ref] < required_amount)
-					continue
-				if (R.is_match(P.material))
-					mats_used[R] = P_ref
-					mats_projected[P_ref] -= required_amount
-					break
+			var/piece_ref = src.get_material_for_requirement(R, M.item_requirements[R], mats_reserved)
+			if (!isnull(piece_ref))
+				mats_used[R] = piece_ref
 
 		return mats_used
 
+	/// Get the material in storage which satisfies some amount of a requirement.
+	proc/get_material_for_requirement(datum/manufacturing_requirement/R, var/required_amount, var/list/mats_reserved)
+		var/list/C = src.get_contents()
+		for (var/obj/item/material_piece/P as anything in C)
+			if (!R.is_match(P.material))
+				continue
+			// We can use this material! Get the amount of free material and reserve/mark as used whatever is free.
+			var/P_ref = "\ref[P]"
+			var/amount_free = P.amount - mats_reserved[P_ref]
+			var/amount_to_use = min(amount_free, required_amount / 10)
+			if ((amount_to_use * 10) < required_amount)
+				continue
+			mats_reserved[P_ref] ||= 0
+			mats_reserved[P_ref] += amount_to_use
```

Comment:
Per "thanks BYOND" comment, possible we need to handle dumb additive behavior here as well?

Reply:
Thankfully not, the only way this happens is when maths isn't done right. Rounding here would imply that there's flexibility with item amounts, which, fine, if that's desired -- but it also means people can start doing stuff like
produce 0.1 sheet using 0.990 extremely valuable metal
x10
reclaim sheet
repeat ad infinitum
and now you have infinity infused gibbis material.
IDK, i don't really care about that weird thing. But I feel I have to mention it before someone else does, because it tends to come up.

Reply:
I did try a few implementations here, by the way -- and I settled on the above because any implementation that's not going to address the cause (that 5 - 4.8 is not evaluating to 0.2 as per intent) doesn't stop it from escalating to someone putting in like, 100 bars, printing 1000 sheets, and then figuring out that floating point math makes that last 0.1 unprintable.

If it's just about using both as a sort of layered approach though, sounds fine? As it is now, though, this issue should not come up.